### PR TITLE
#14336: add bfloat8b support for _like creation ops

### PR DESCRIPTION
**Pull request by KalaivaniMCW**
### Ticket
Link to Github Issue #14336
(also addresses issues like #13927 ,  #13775 , #14137 etc )

### Problem description
a. currently, creation ops rely on `tt::numpy` and host operation and it doesn't support `bfloat8_b`, but supports ROW_MAJOR and int dtypes
b. due to this, any eltwise composite op/backward op using creation ops are not supporting `bfloat8_b` dtype.

### What's changed
- modified full_like/zeros_like/ones_like creation ops so that they use device operation `ttnn.fill` for the supported criteria (i.e. TILE layout and bf16/bf8_b/fp32 dtypes) and rely on current implementation for other cases 
- this will provide `bfloat8_b` support to eltwise ops like softshrink_bw and address a list of GH issues 

**Profiling report on main:**
**op,count,python min dispatch time (ms),python mean dispatch time(ms),python mean dispatch + sync time (ms)**
ttnn.full_like,800,**0.779**,0.787,0.799
ttnn.ones_like,800,**0.783**,0.81,0.816
ttnn.zeros_like,800,**0.8**,0.805,0.821

vs 

**Profiling report on this branch:** 
**op,count,python min dispatch time (ms),python mean dispatch time(ms),python mean dispatch + sync time (ms),C++ mean dispatch time (ms)**
ttnn.full_like,800,**0.038**,0.039,0.101,0.012
ttnn.ones_like,800,**0.03**,0.032,0.101,0.012
ttnn.zeros_like,800,**0.03**,0.033,0.101,0.012

a few dependant ops:
on main: 
ttnn.celu_bw,800,1.986,2.065,2.468,0.135
ttnn.hardsigmoid,800,1.828,1.845,1.873,0.041
ttnn.hardtanh_bw,800,2.039,2.088,2.631,0.163
ttnn.softshrink_bw,800,2.721,2.841,2.891,0.114
ttnn.xlogy,800,1.194,1.206,1.833,0.122

vs 

ttnn.celu_bw,800,0.416,0.453,1.451,0.158
ttnn.hardsigmoid,800,0.201,0.214,0.574,0.064
ttnn.hardtanh_bw,800,0.539,0.594,1.626,0.203
ttnn.softshrink_bw,800,0.403,0.425,1.322,0.144
ttnn.xlogy,800,0.41,0.441,1.318,0.148

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11566088634
https://github.com/tenstorrent/tt-metal/actions/runs/11593065696
https://github.com/tenstorrent/tt-metal/actions/runs/11677532478
- [x] Nightly FD https://github.com/tenstorrent/tt-metal/actions/runs/11554295346
same as main https://github.com/tenstorrent/tt-metal/actions/runs/11659568349
https://github.com/tenstorrent/tt-metal/actions/runs/11677565314
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by umadevimcw — APPROVED** [review]

**Review by mouliraj-mcw — APPROVED** [review]

**Review by VirdhatchaniKN — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp [hunk 1]
```diff
@@ -30,8 +30,8 @@ void validate_supported_arch_dtype(
         case UnaryOpType::FILL:
            if(arch == tt::ARCH::GRAYSKULL){
                 TT_FATAL(
-                    input_datatype == DataType::BFLOAT16,
-                    "Supported input data type '{}' for UnaryOpType '{}' on Grayskull.",
+                    (input_datatype == DataType::BFLOAT16 || input_datatype == DataType::BFLOAT8_B),
+                    "Unsupported input data type '{}' for UnaryOpType '{}' on Grayskull.",
                     static_cast<int>(input_datatype),
                     static_cast<int>(op_type));
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp:36]: static_cast is necessary here?

**Review by eyonland — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp (hunk 1 above)
Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp:36]: all the validation checks use this pattern here

**Review by Aswinmcw — APPROVED** [review]
LGTM

File: ttnn/cpp/ttnn/operations/creation.hpp [hunk 2]
```diff
@@ -138,23 +139,34 @@ inline ttnn::Tensor full_like_impl(
...
-            layout.value_or(tensor.get_layout()),
-            device.value_or(*tensor.device()),
-            memory_config.value_or(tensor.memory_config()),
-            optional_output_tensor);
+        // requires reference tensor to be in TILE for device operation fill - this will be changed later
+        if (is_TILE &&  (dtype_value == DataType::BFLOAT8_B || dtype_value == DataType::BFLOAT16 || (arch != tt::ARCH::GRAYSKULL && dtype_value == DataType::FLOAT32)) && tensor.storage_type() == StorageType::DEVICE ) {
+            return ttnn::fill(tensor, fill_value, memory_config, optional_output_tensor);
+        }
+        else {
+            return full_impl(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/creation.hpp:153]: This `full` is on host, right?
There is also moreh_full - on device, but I'd guess it does not support bf8_b?

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp [hunk 3]
```diff
@@ -30,8 +30,8 @@ void validate_supported_arch_dtype(
         case UnaryOpType::FILL:
            if(arch == tt::ARCH::GRAYSKULL){
                 TT_FATAL(
-                    input_datatype == DataType::BFLOAT16,
-                    "Supported input data type '{}' for UnaryOpType '{}' on Grayskull.",
+                    (input_datatype == DataType::BFLOAT16 || input_datatype == DataType::BFLOAT8_B),
+                    "Unsupported input data type '{}' for UnaryOpType '{}' on Grayskull.",
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp:34]: Let's tell a user what types are supported

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp:34]: also, does it mean we support any type on wh+?

File: tests/ttnn/unit_tests/operations/test_creation.py [hunk 4]
```diff
@@ -115,6 +158,32 @@ def test_full_like(device, input_shape, fill_value):
     assert torch.allclose(torch_output_tensor, output_tensor)
 
 
+@pytest.mark.parametrize(
+    "input_shape",
+    [
+        [32, 32],
+        [5, 96, 64],
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_creation.py:165]: Lets test shapes in rank 2-5 today.

Later, once tensor layout work is complete - we will have to start testing with shapes of rank 0 to 5 and additionally for 0 volume.

File: tests/ttnn/unit_tests/operations/eltwise/test_composite.py [hunk 5]
```diff
@@ -235,6 +239,25 @@ def test_unary_composite_hardsigmoid_ttnn(input_shapes, device):
...
 
 
+@pytest.mark.parametrize(
+    "input_shapes",
+    (
+        (torch.Size([1, 1, 32, 32])),
+        (torch.Size([1, 1, 320, 384])),
+        (torch.Size([1, 3, 320, 384])),
+    ),
+)
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/eltwise/test_composite.py:249]: For consistency, maybe we can use the same format and skip `torch.Size` part?

```@pytest.mark.parametrize(
    "input_shape",
    [
        [32, 32],
        [5, 96, 64],
    ],
)
```

File: tests/ttnn/unit_tests/operations/eltwise/test_composite.py [hunk 6]
```diff
@@ -235,6 +239,25 @@ def test_unary_composite_hardsigmoid_ttnn(input_shapes, device):
...
+def test_unary_composite_hardsigmoid_bf8b(input_shapes, device):
+    in_data1, input_tensor1 = data_gen_with_range_dtype(input_shapes, -100, 100, device, False, False, ttnn.bfloat8_b)
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/eltwise/test_composite.py:251]: I am curious, would we generate a test for many different datatypes, making the last parameter `ttnn.bfloat8_b` a test param?

File: ttnn/cpp/ttnn/operations/creation.hpp (hunk 2 above)
Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/creation.hpp:153]: Yes. full_impl is on host. That's right, moreh_full does not support bf8_b.

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp (hunk 3 above)
Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp:34]: for WH, we support BFLOAT16, BFLOAT8_B, FLOAT_32. for other dtypes, we'll rely on current implementation for now

File: tests/ttnn/unit_tests/operations/eltwise/test_composite.py (hunk 6 above)
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_composite.py:251]: yes. by default `data_gen_with_range` uses dtype=ttnn.bfloat16, we use `data_gen_with_range_dtype` when there needs to be specific dtype tested (usually for a GH issue)

File: tests/ttnn/unit_tests/operations/eltwise/test_composite.py (hunk 5 above)
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/eltwise/test_composite.py]: @ayerofieiev-tt yes, done.

File: tests/ttnn/unit_tests/operations/test_creation.py (hunk 4 above)
Comment (KalaivaniMCW) [reply tests/ttnn/unit_tests/operations/test_creation.py]: @ayerofieiev-tt yeah sure. I've added shapes with rank 2-5

File: tests/ttnn/unit_tests/operations/test_creation.py [hunk 7]
```diff
@@ -32,6 +32,27 @@ def test_zeros_like(device, input_shape):
...
+)
+def test_zeros_like_bf8b(device, input_shape):
+    torch_input_tensor = torch.rand((input_shape), dtype=torch.bfloat16)
+    torch_output_tensor = torch.zeros_like(torch_input_tensor)
+
+    input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat8_b, layout=ttnn.TILE_LAYOUT, device=device)
+    output_tensor = ttnn.zeros_like(input_tensor)
+    assert ttnn.is_tensor_storage_on_device(output_tensor)
+    output_tensor = ttnn.from_device(output_tensor)
+    output_tensor = ttnn.to_torch(output_tensor).to(torch.bfloat16)
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/operations/test_creation.py:50]: can this be a single op call `ttnn.to_torch`?

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp [hunk 8]
```diff
@@ -30,10 +30,8 @@ void validate_supported_arch_dtype(
         case UnaryOpType::FILL:
            if(arch == tt::ARCH::GRAYSKULL){
                 TT_FATAL(
-                    input_datatype == DataType::BFLOAT16,
-                    "Supported input data type '{}' for UnaryOpType '{}' on Grayskull.",
-                    static_cast<int>(input_datatype),
-                    static_cast<int>(op_type));
+                    (input_datatype == DataType::BFLOAT16 || input_datatype == DataType::BFLOAT8_B),
+                    "Supported dtypes on Grayskull are BFLOAT16/BFLOAT8_B.");
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_device_operation.cpp:34]: https://github.com/tenstorrent/tt-metal/blob/main/ErrorMessageBestPractices.md

"Unsupported dtype {}. On Grayskull only BFLOAT16/BFLOAT8_B are supported"

Inclusion of the actual type will be useful for users

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**